T2D-208 Secure metadata APIs (#57)
EBIvariation/metadata-ws: docs/security.md

   - Resource server: Metadata Application
   - Resource owner: Keycloak users with allowed role
   - Client: Application calling Metadata REST APIs

No restrictions are applied on the `/` and `/swagger-ui.html` endpoints, which can be accessed by anonymous users, for example via http://localhost:8085/ and http://localhost:8085/swagger-ui.html. Among all other endpoints, read-only (GET) operation can be issued by anonymous users and update operations (POST, PUT, PATCH, DELETE) can be performed only by users with the "SERVICE_OPERATOR" role (described below).

## Setup

   - Enter the Keycloak public key value in the `application.properties` file, property _security.oauth2.resource.jwt.key-value_. It should have a value similar to `-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9wxxxxxxxxxxxxxxxxxxxffjZs/D+e9A56XuRJSQ9QIDAQAB\n-----END PUBLIC KEY-----`
   - This can be obtained from Keycloak Realm settings -> Keys -> Public key, at the right-most column.
  
 
## Configuring Keycloak
  
1. Setup and configure Keycloak server with realm, client, role, mapper and user [as described in the official documentation](https://www.keycloak.org/docs/latest/getting_started/index.html)

> 2. Consider following are configured in Keycloak as an example

Reviewer: ```suggestion
2. Ensure that the following Keycloak settings are configured as described:
```

>      - Role: �SERVICE_OPERATOR�. This role name is mandatory. It could be either client role or realm role.

Reviewer: ```suggestion
     - Role "SERVICE_OPERATOR": This role name is mandatory. It could be either client role or realm role.
```

Author: Again `Add suggestion to branch` was grayed. Clicking `Commit suggestion` made all as outdated and didn't allow further Commit suggestion. Did manual changes!

>      - User name Mapper: Keycloak returns access token in JWT. Logged in username is returned in claim �preferred\_username� in access token. Spring Security OAuth2 resource Server expects username in claim �user\_name�. Hence a mapper to map logged in username to a new claim named user_name is required

Reviewer: ```suggestion
     - User name Mapper: Keycloak returns access token in JWT. The logged in username is returned in claim "preferred\_username" in said access token, but Spring Security OAuth2 expects the username in claim "user\_name". Hence a mapper to a new claim named "user\_name" is required.
```

# Security

> This application uses Spring OAuth2 and Keycloak authorization to secure its REST APIs. The mappings from OAuth2 terminology to our implementation are:

Author: Would it be better to restrict the characters/line; view does not depend on editor width. Or view as per the editor width without character/line restriction. 
Please let me what is suggested

> This application uses Spring OAuth2 and Keycloak authorization to secure its REST APIs. The mappings from OAuth2 terminology to our implementation are:

Reviewer: Following the style of other README files, we don't limit the line width to 120 characters. Every paragraph is grouped in a single line.

   - Authorization server: Keycloak server
   - Resource server: Metadata Application
   - Resource owner: Keycloak users with allowed role
   - Client: Application calling Metadata REST APIs

Endpoints

- No restrictions are applied on the `/` and `/swagger-ui.html` endpoints, which can be accessed by anonymous users, for example via http://localhost:8085/ and http://localhost:8085/swagger-ui.html
- Among all other endpoints:

>      - Read-only (GET) operations can be issued by any user authenticated with Keycloak. The user can have any role assigned.

Author: https://github.com/EBIvariation/metadata-ws/pull/57#discussion_r299783854

     - Update operations (POST, PUT, PATCH, DELETE) can be performed only by users with the "SERVICE_OPERATOR" role (described below).

##Setup

   - Enter the Keycloak public key value in the `application.properties` file, property _security.oauth2.resource.jwt.key-value_. It should have a value similar to `-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9wxxxxxxxxxxxxxxxxxxxffjZs/D+e9A56XuRJSQ9QIDAQAB\n-----END PUBLIC KEY-----`
   - This can be obtained from Keycloak Realm settings -> Keys -> Public key, at the right-most column.
  
 
##Configuring keycloak
  

> 1. Setup and configure keycloak server with realm, client, role, mapper and user [as described in the official documentation](https://www.keycloak.org/docs/latest/getting_started/index.html)

Author: I will change to K wherever k is used, except link `https://www.keycloak.org/docs/latest/getting_started/index.html`

## Security

Metadata application uses Spring OAuth2 and Keycloak authorization to secure REST APIs. W.r.t OAuth2 standards mappings are
   - Authorization server: Keycloak server
   - Resource server: Metadata Application
   - Resource owner: Keycloak users with allowed role
   - Client: Application calling Metadata REST APIs

> Endpoints

Author: Didnt get. I was describing endpoints

> Endpoints

Reviewer: I just didn't think it was worth creating a subsection with bullet points. I find easier to read if the heading is removed, and [the suggestion](https://github.com/EBIvariation/metadata-ws/pull/57#discussion_r299500699) to convert to regular text is applied.

> Endpoints

Reviewer 2: Ah I get it now! @iyangar Could you please remove all bullets in the “Endpoints” section, and also remove the “Endpoints” header as well?

   - / and /swagger-ui.html can be accessed by ananymous user. Ex: http://localhost:8085/ and http://localhost:8085/swagger-ui.html
   - All other endpoints: 

>         - Read(GET) operation can be issued by any authenticated user with keycloak. The user can have any role assigned

Author: Initial spaces to indent

>         - Read(GET) operation can be issued by any authenticated user with keycloak. The user can have any role assigned

Reviewer: Right now, given that we don't provide access to studies previous to publication, read operations should be anonymous instead of via a read-only user.

        - Update operation (POST, PUT, PATCH, DELETE) can be done by users only with “SERVICE_OPERATOR” role(described later)

**Building metadata API application**

   - Enter keycloak public key value in Application.properties “security.oauth2.resource.jwt.key-value”. It should have a value like “-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9wxxxxxxxxxxxxxxxxxxxffjZs/D+e9A56XuRJSQ9QIDAQAB\n-----END PUBLIC KEY-----“
   - This can be obtained from Keycloak Realm settings->Keys->Public key at the right-most column
  
 

> **Configuring keycloak**

Reviewer 2: @iyangar This is not fixed yet (missing space, Keycloak with capital K)

> **Configuring keycloak**

Reviewer: Suggestions are easily applied by clicking on "add suggestion to batch", then committing them altogether. That way the formatting of this and other of my suggestions will be applied exactly as I typed it.

> **Configuring keycloak**

Author: "add suggestion to batch" is not enabled for me(may be outdated comment) and surprisingly for latest suggested change also https://github.com/EBIvariation/metadata-ws/pull/57#discussion_r302084070

  
1. Setup and configure keycloak server with realm, client, role, mapper and user

> 2. Consider following are configured in keycloak as an example

Author: Sure, please let me which lines i can remove

> 2. Consider following are configured in keycloak as an example

Author: This part is done and only necessary lines are kept. If there are further comments i will address

> 2. Consider following are configured in keycloak as an example

Reviewer: In fact, I suggested to keep only "role" and "user name mapper".

     - Keycloak server running at: “http://localhost:8080” 
     - Relam: “securemetadata”
     - Client id: “secure-client”
     - Role: “SERVICE_OPERATOR”. This role name is mandatory. It could be either client role or realm role.
     - User name Mapper: Keycloak returns access token in JWT. Logged in username is returned in claim “preferred\_username” in access token. Spring Security OAuth2 resource Server expects username in claim “user\_name”. Hence a mapper to map logged in username to a new claim named user_name is required
     - User: “svc_opr”

>      - Redirect URL in client settings: “http://localhost:8085” (Metadata application)

Author: In keycloak create a client and in its settings mention the URL where metadata application is running.

>      - Redirect URL in client settings: “http://localhost:8085” (Metadata application)

Reviewer: Then this is badly explained. I suggest the following
```suggestion
     - 'Redirect URL' in client settings should point to where the metadata API is running
```

> 3. Get all security endpoints from OpenID configuration URL "http://localhost:8080/auth/realms/securemetadata/.well-known/openid-configuration". 

Author: Keycloak exposes this `<server>/auth/realms/<realm>/.well-known/openid-configuration` . I have removed it.

      - It should contain several endpoints and other information, which should appear like
      >"issuer": "http://localhost:8080/ega/ampt2d/auth/realms/securemetadata" <br>
      >“authorization_endpoint": "http://localhost:8080/ega/ampt2d/auth/realms/securemetadata/protocol/openid-connect/auth" <br>
      >"token_endpoint": "http://localhost:8080/ega/ampt2d/auth/realms/securemetadata/protocol/openid-connect/token" <br>
      >"token\_introspection_endpoint": "http://localhost:8080/ega/ampt2d/auth/realms/securemetadata/protocol/openid-connect/token/introspect" <br>
      >"userinfo_endpoint": "http://localhost:8080/ega/ampt2d/auth/realms/securemetadata/protocol/openid-connect/userinfo"

   

> **Get authorization token from keycloak server using postman**

Reviewer: Again I think this is too much information, and the URLs are misleading because this service won't be deployed for the EGA in a while. Maybe this could be turned into a couple of points like the following:
```suggestion

## Testing Keycloak
1. Get authorization token from the Keycloak server using Postman (for testing)
2. In order to send the token
2.1. From Swagger: click the `Authorize` button and enter "Bearer <token obtained>"
2.2. From a REST client: include header "Authorization" with value "Bearer <token obtained>"
```
Everything below this point can then be dropped. If people experience trouble there are plenty of pages to check literally searching for "Get authorization token from keycloak server using postman".

Reviewer 3: I agree with @cyenyxe. I guess Readme can contain only how to get authorization and how to test the application like provided in this. eg: https://github.com/EBIvariation/metadata-ws/blob/keycloakOauth/README.md. All other info regarding the setup of Keycloak can be put in confluence or the ticket.

Author: Should the section heading be `Testing keycloak`?
